setup.py

```diff
@@ -21,7 +36,11 @@
     entry_points={
         'redash.extensions': [
             'dockerflow = redash_stmo.dockerflow:dockerflow',
-            'datasource_health = redash_stmo.health:datasource_health'
+            'datasource_health = redash_stmo.health:datasource_health',
+            'datasource_link = redash_stmo.datasource_link:datasource_link'
+        ],
+        'webpack.bundles': [
+            'datasource_link = redash_stmo.js_extensions:datasource_link',
```

Comment:
So looking at this I noticed a problem I hadn't seen when I reviewed the extension API code into our fork, that the need to provide config values for `extension_directory` and `frontend_content` isn't actually really there and looks like redundant. Both values can be calculated from the entrypoint object that is returned when we load the `redash.extensions` entrypoints.

- `extension_directory` -- is the filesystem path of the package `redash_stmo.datasource_link`
- `frontend_content` -- has no need to be defined individually per extension and should instead be consistent for every extension

So I believe we should simplify the structure of an extension further:

#### An extension is a Python module or package which may or may not have a `bundle` subdirectory, and is defined in the `redash.extensions` entrypoint.

The code that loads the bundles in [bundle-extensions](https://github.com/mozilla/redash/blob/9424dadcc7c0e13d9f094b18914615bae756dd92/bin/bundle-extensions#L25-L30) should calculate the original source of the extension frontend files from the entry in the `redash.extensions` entrypoint and use the name of the loaded entrypoint (e.g. `datasource_link`) as the subdirectory in `client/app/extensions`.

That basically would remove the need for the `webpack.bundles` entrypoint again and consolidate frontend and backend into one location.

Reply:
What about when an extension is frontend only with no python portion? Would we then have an empty python file/function for it to be loaded here: https://github.com/mozilla/redash/blob/master/redash/extensions.py?

Alternatively, `extensions.py` could be modified to skip over empty extensions. 

I think this is originally why I had a separate entry, `webpack.bundles` here. Since not all frontend extensions will have a backend portion.

Reply:
For pure-frontend extensions the process would still be having a package with its name, e.g.

```
src
└── redash_stmo
    ├── __init__.py
    ├── dockerflow.py
    ├── health.py
    ├── whimsycorn
    │   ├── __init__.py
    │   └── bundle
    │   │   ├── index.js
    │   │   ├── style.css
    │   │   └── template.html
    ├── query_runner
    │   ├── __init__.py
    │   └── activedata.py
    └── settings.py
```

That would allow us:

a) use the Python import system to look for the path of the packages

b) provide a location of extra extension metadata later on (e.g. verbose name, description text etc)

Skipping over extensions by checking if the loaded entrypoint is a callable would make sense. IIRC the Entrypoint object has some good APIs to see what it is.

E.g. the syntax for entrypoints only requires a modulename (read: dotted path to a module or package): https://github.com/pypa/setuptools/blob/01fd232d95ad51f784de53c0993a7b91868d269d/pkg_resources/__init__.py#L2361-L2368

You can use the [`Entrypoint.resolve`](https://github.com/pypa/setuptools/blob/01fd232d95ad51f784de53c0993a7b91868d269d/pkg_resources/__init__.py#L2327-L2335) function which will try to import it and raise an `ImportError` if it fails when you skip those that don't actually have an entrypoint provided.

So an entrypoint list in setup.py could look like this (where the whimsycorn extension just happens to not need a callable):

```python
entrypoints={
    'redash.extensions': [
        'dockerflow = redash_stmo.dockerflow:dockerflow',
        'datasource_health = redash_stmo.health:datasource_health',
        'datasource_link = redash_stmo.datasource_link:datasource_link',
        'whimsycorn = redash_stmo.whimsy_corn',
   ], 
}
```

In the bundle-extensions script you'll repeat some of the logic of course since you'll need to check if the bundle directory exists in the extension directory.